cli_tools/gce_ovf_import/main.go

```diff
@@ -0,0 +1,501 @@
+	instance.CanIpForward = *canIPForward
+	instance.DeletionProtection = *deletionProtection
+	if instance.Scheduling == nil {
+		instance.Scheduling = &compute.Scheduling{}
+	}
+	if *noRestartOnFailure {
+		vFalse := false
+		instance.Scheduling.AutomaticRestart = &vFalse
+	}
+}
+
+func toWorkingDir(dir string) string {
+	wd, err := filepath.Abs(filepath.Dir(*currentExecutablePath))
+	if err == nil {
+		return path.Join(wd, dir)
+	}
+	return dir
+}
+
+// creates a new Daisy Compute client
+func createComputeClient(ctx *context.Context) (daisycompute.Client, error) {
+	computeOptions := []option.ClientOption{option.WithCredentialsFile(*oauth)}
+	if *ce != "" {
+		computeOptions = append(computeOptions, option.WithEndpoint(*ce))
+	}
+
+	computeClient, err := daisycompute.NewClient(*ctx, computeOptions...)
+	if err != nil {
+		return nil, err
+	}
+	return computeClient, nil
+}
+
+// OVFImporter is responsible for importing OVF into GCE
+type OVFImporter struct {
+	ctx                   context.Context
+	storageClient         commondomain.StorageClientInterface
+	computeClient         daisycompute.Client
+	tarGcsExtractor       commondomain.TarGcsExtractorInterface
+	mgce                  commondomain.MetadataGCEInterface
+	ovfDescriptorLoader   domain.OvfDescriptorLoaderInterface
+	bucketIteratorCreator commondomain.BucketIteratorCreatorInterface
+	logger                logging.LoggerInterface
+	zoneValidator         commondomain.ZoneValidatorInterface
+	gcsPathToClean        string
+	workflowPath          string
+	buildID               string
+	diskInfos             *[]ovfutils.DiskInfo
+}
+
+// NewOVFImporter creates an OVF importer, including automatically poulating dependencies,
+// such as compute/storage clients.
+func NewOVFImporter() (*OVFImporter, error) {
+	ctx := context.Background()
+	sc, err := storage.NewClient(ctx)
+	logger := logging.NewLogger("[import-ovf]")
+	if err != nil {
+		return nil, err
+	}
+	storageClient, err := storageutils.NewStorageClient(ctx, sc, logger)
+	if err != nil {
+		return nil, err
+	}
+	computeClient, err := createComputeClient(&ctx)
+	if err != nil {
+		return nil, err
+	}
+	tarGcsExtractor := storageutils.NewTarGcsExtractor(ctx, storageClient, logger)
+	buildID := os.Getenv("BUILD_ID")
+	if buildID == "" {
+		buildID = utils.RandString(5)
+	}
+	workingDirOVFImportWorkflow := toWorkingDir(ovfImportWorkflow)
+	bic := &storageutils.BucketIteratorCreator{}
+
+	ovfImporter := &OVFImporter{ctx: ctx, storageClient: storageClient, computeClient: computeClient,
+		tarGcsExtractor: tarGcsExtractor, workflowPath: workingDirOVFImportWorkflow, buildID: buildID,
+		ovfDescriptorLoader: ovfutils.NewOvfDescriptorLoader(storageClient),
+		mgce:                &computeutils.MetadataGCE{}, bucketIteratorCreator: bic, logger: logger,
+		zoneValidator: &computeutils.ZoneValidator{ComputeClient: computeClient}}
+	return ovfImporter, nil
+}
+
+func (oi *OVFImporter) populateMissingParameters() error {
+	aProject := *project
```

Comment:
Its generally considered not a great idea to modify parsed flag values, or even refer to flag values in most functions besides main. Instead pass args around, this simplifies testing and eliminates any potential race conditions if you happen to use concurrency later on.

You can keep the flag reference in this function if you want (maybe call it parseFlags or something), but then this should return the parsed values that you then pass to other functions.

Reply:
Refactored